File: lib/screens/category_detail_screen.dart
import 'package:flutter/material.dart';
import 'add_transaction_screen.dart';
import 'package:isar/isar.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:provider/provider.dart';
import '../services/auth_service.dart';
import 'package:myapp/main.dart';
import 'package:myapp/models/transaction_model.dart';
import 'package:myapp/screens/home.dart';

class Transaction {
  final IconData icon;
  final Color iconColor;
  final String title;
  final String date;
  final String category;
  final double amount;
  final String type;

  Transaction({
    required this.icon,
    required this.iconColor,
    required this.title,
    required this.date,
    required this.category,
    required this.amount,
    required this.type,
  });
}

class CategoryDetailScreen extends StatefulWidget {
  final String categoryName;
  const CategoryDetailScreen({super.key, required this.categoryName});

  @override
  State<CategoryDetailScreen> createState() => _CategoryDetailScreenState();
}

class _CategoryDetailScreenState extends State<CategoryDetailScreen> {
  List<String> months = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];

  List<Transaction> transactions = [];

  Future<void> _getData(String uid) async {
    final List<TransactionModel> query = await isar.transactionModels.filter().categoryEqualTo(this.widget.categoryName).findAll();
    List<Transaction> temp = [];
    if (query.isNotEmpty) {
      for (final transaction in query) {
        temp.add(
          Transaction(
            icon: Icons.fastfood_outlined,
            iconColor: Colors.red.shade300,
            title: transaction.title,
            date: "${months[transaction.timestamp.month - 1]} ${transaction.timestamp.day}",
            category: transaction.category,
            amount: transaction.amount,
            type: transaction.type,
          ),
        );
      }
      setState(() {
        transactions = temp;
      });
    } else {
            if (mounted) {
        showDialog(
          context: context,
          barrierDismissible: false,
          builder: (BuildContext dialogContext) {
            return AlertDialog(
              title: const Text('No Transactions'),
              content: const Text("You haven't added any transactions yet."),
              actions: <Widget>[
                TextButton(
                  child: const Text('OK'),
                  onPressed: () {
                    Navigator.of(dialogContext).pop(); // Dismiss dialog
                    Navigator.pushReplacement(context, MaterialPageRoute(builder: (context) => HomePage()));
                  },
                ),
              ],
            );
          },
        );
      }
    }
  }

  User? _currentUser;

  @override
  void initState() {
    super.initState();
        WidgetsBinding.instance.addPostFrameCallback((_) {
      _currentUser = Provider.of<AuthService>(
        context,
        listen: false,
      ).currentUser;
      if (_currentUser != null) {
        _getData(_currentUser!.uid);
      } else {
        // Handle case where user is not logged in, maybe navigate to login or show a message
        print("User not logged in.");
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    // Dummy data for transactions

    return Scaffold(
      appBar: AppBar(
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          onPressed: () {
            Navigator.pop(context);
          },
        ),
        title: Text(this.widget.categoryName),
        actions: [
          IconButton(
            icon: const Icon(Icons.notifications_none),
            onPressed: () {
              // Handle notification icon tap
            },
          ),
        ],
        backgroundColor: const Color(
          0xFFC8E6C9,
        ), // Adjust color to match design
        elevation: 0,
      ),
      body: Column(
        children: [
          Container(
            padding: const EdgeInsets.all(16.0),
            color: const Color(0xFFC8E6C9), // Adjust color to match design
            child: Column(
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: const [
                        Text(
                          'Total Balance',
                          style: TextStyle(fontSize: 16, color: Colors.black54),
                        ),
                        Text(
                          '\$7,783.00',
                          style: TextStyle(
                            fontSize: 24,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                      ],
                    ),
                    Column(
                      crossAxisAlignment: CrossAxisAlignment.end,
                      children: const [
                        Text(
                          'Total Expense',
                          style: TextStyle(fontSize: 16, color: Colors.black54),
                        ),
                        Text(
                          '-\$1,187.40',
                          style: TextStyle(
                            fontSize: 24,
                            fontWeight: FontWeight.bold,
                            color: Colors.redAccent,
                          ),
                        ),
                      ],
                    ),
                  ],
                ),
                const SizedBox(height: 16.0),
                LinearProgressIndicator(
                  value: 0.3, // Example progress value
                  backgroundColor: Colors.grey[300],
                  valueColor: const AlwaysStoppedAnimation<Color>(Colors.green),
                ),
                const SizedBox(height: 8.0),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: const [
                    Text('30% Of Your Expenses, Looks Good.'),
                    Text('\$20,000.00'),
                  ],
                ),
              ],
            ),
          ),
          Expanded(
            child: ListView.builder(
              itemCount: transactions.length,
              itemBuilder: (context, index) {
                final transaction = transactions[index];
                return ListTile(
                  leading: const Icon(
                    Icons.fastfood,
                  ), // Replace with appropriate icon based on category/type
                  title: Text(transaction.title),
                  subtitle: Text(transaction.date),
                  trailing: Text(
                    '${transaction.type != "expense" ? '+' : '-'}\$${transaction.amount.abs().toStringAsFixed(2)}',
                    style: TextStyle(
                      color: transaction.type != "expense"
                          ? Colors.green
                          : Colors.redAccent,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                );
              },
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(16.0),
            child: ElevatedButton(
              onPressed: () {
                // Navigate to Add Transaction screen
                Navigator.push(
                  context,
                  MaterialPageRoute(
                    builder: (context) => AddTransactionScreen(
                      categoryName: this.widget.categoryName,
                    ),
                  ),
                );
              },
              style: ElevatedButton.styleFrom(
                backgroundColor: Colors.green, // Adjust color to match design
                padding: const EdgeInsets.symmetric(vertical: 16.0),
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(8.0),
                ),
              ),
              child: Center(
                child: const Text(
                  'Add Expenses', // Or 'Add Income' depending on context
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

File: src/screens/category_detail_screen.js
import React, 'react';
import {
  View,
  Text,
  StyleSheet,
  FlatList,
  TouchableOpacity,
  Alert,
} from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Ionicons } from '@expo/vector-icons';
import { useFocusEffect } from '@react-navigation/native';
import auth from '@react-native-firebase/auth';

const months = [
  'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December'
];

const Header = () => (
  <View style={styles.headerContainer}>
    <View style={styles.balanceRow}>
      <View>
        <Text style={styles.balanceLabel}>Total Balance</Text>
        <Text style={styles.balanceValue}>$7,783.00</Text>
      </View>
      <View style={{ alignItems: 'flex-end' }}>
        <Text style={styles.balanceLabel}>Total Expense</Text>
        <Text style={[styles.balanceValue, { color: 'red' }]}>-$1,187.40</Text>
      </View>
    </View>
    <View style={styles.progressContainer}>
      <View style={styles.progressBar} />
    </View>
    <View style={styles.progressLabelContainer}>
        <Text>30% Of Your Expenses, Looks Good.</Text>
        <Text>$20,000.00</Text>
    </View>
  </View>
);

const TransactionItem = ({ item }) => {
    const date = new Date(item.timestamp);
    const formattedDate = `${months[date.getMonth()]} ${date.getDate()}`;
    const amountColor = item.type === 'expense' ? 'red' : 'green';
    const sign = item.type === 'expense' ? '-' : '+';
    // A simple mapping from category to icon
    const iconName = item.category === 'Food' ? 'fast-food' : 'apps';

    return (
        <View style={styles.transactionItem}>
            <Ionicons name={iconName} size={24} color="black" style={styles.transactionIcon} />
            <View style={styles.transactionDetails}>
                <Text style={styles.transactionTitle}>{item.title}</Text>
                <Text style={styles.transactionDate}>{formattedDate}</Text>
            </View>
            <Text style={[styles.transactionAmount, { color: amountColor }]}>
            {`${sign}$${Math.abs(item.amount).toFixed(2)}`}
            </Text>
        </View>
    );
};


const CategoryDetailScreen = ({ route, navigation }) => {
  const { categoryName } = route.params;
  const [transactions, setTransactions] = React.useState([]);
  const [currentUser, setCurrentUser] = React.useState(null);


  useFocusEffect(
    React.useCallback(() => {
      const getData = async () => {
        const storedTransactions = await AsyncStorage.getItem('transactions');
        if (storedTransactions) {
          const allTransactions = JSON.parse(storedTransactions);
          const filtered = allTransactions.filter(
            (t) => t.category === categoryName
          );

          if (filtered.length > 0) {
            setTransactions(filtered);
          } else {
            Alert.alert(
              'No Transactions',
              "You haven't added any transactions for this category yet.",
              [{ text: 'OK', onPress: () => navigation.goBack() }]
            );
          }
        } else {
          Alert.alert(
            'No Transactions',
            "You haven't added any transactions for this category yet.",
            [{ text: 'OK', onPress: () => navigation.goBack() }]
          );
        }
      };

      const user = auth().currentUser;
      if (user) {
        setCurrentUser(user);
        getData();
      } else {
        console.log('User not logged in');
        // Optionally navigate to login screen
      }
    }, [categoryName, navigation])
  );


  return (
    <View style={styles.container}>
      <Header />
      <FlatList
        data={transactions}
        renderItem={({ item }) => <TransactionItem item={item} />}
        keyExtractor={(item) => item.id.toString()}
        style={styles.list}
      />
      <TouchableOpacity
        style={styles.addButton}
        onPress={() => navigation.navigate('AddTransaction', { categoryName })}
      >
        <Text style={styles.addButtonText}>Add Expenses</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#C8E6C9',
  },
  headerContainer: {
    padding: 16,
    backgroundColor: '#C8E6C9',
  },
  balanceRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  balanceLabel: {
    fontSize: 16,
    color: 'black',
  },
  balanceValue: {
    fontSize: 24,
    fontWeight: 'bold',
  },
  progressContainer: {
    height: 10,
    backgroundColor: '#BDBDBD',
    borderRadius: 5,
    marginTop: 16,
  },
  progressBar: {
    height: 10,
    width: '30%', // example
    backgroundColor: 'green',
    borderRadius: 5,
  },
  progressLabelContainer: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      marginTop: 8
  },
  list: {
    flex: 1,
    backgroundColor: 'white',
  },
  transactionItem: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 16,
    borderBottomWidth: 1,
    borderBottomColor: '#eee',
  },
  transactionIcon: {
      marginRight: 16,
  },
  transactionDetails: {
      flex: 1,
  },
  transactionTitle: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  transactionDate: {
    fontSize: 14,
    color: 'grey',
  },
  transactionAmount: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  addButton: {
    backgroundColor: 'green',
    padding: 16,
    margin: 16,
    borderRadius: 8,
    alignItems: 'center',
  },
  addButtonText: {
    color: 'white',
    fontSize: 18,
    fontWeight: 'bold',
  },
});

export default CategoryDetailScreen;